### PR TITLE
[stackalloc-init] Symbol API tests

### PR DESCRIPTION
**Pull request by alrz**

### CODE REVIEW

**Comment by alrz** [comment]
We do not have Operation for stackalloc, if the following is acceptable I'll push them here.
```
IOperation:  (OperationKind.None, Type: null, IsInvalid) (Syntax: 'stackalloc int[] { 1 }')
  Children(1):
      ILiteralOperation (OperationKind.Literal, Type: System.Int32, Constant: 1, IsInvalid, IsImplicit) (Syntax: 'stackalloc int[] { 1 }')
```

**Comment by jcouv** [comment]
Thanks. I think the main concern was to check that `GetTypeInfo` and `GetSymbolInfo` should work on the dimension specifier and within the initializers, and that the initializer elements should be listed as children from the IOperation.

**Comment by alrz** [comment]
Note: `FlowAnalysisTests.TestStackAllocArrayInitializer` already covers both read and write.

File: src/Compilers/CSharp/Test/Semantic/Semantics/StackAllocInitializerTests.cs [hunk 1]
```diff
@@ -2014,5 +1950,147 @@ static int M(System.Action action)
...
+        public void StackAllocArrayCreationExpression_Symbols()
+        {
+            var comp = CreateCompilationWithMscorlibAndSpan(@"
+using System;
+unsafe class Test
+{
+    public void Method1()
+    {
+        var obj1 = stackalloc double[2] { 1, 1.2 };
+        Span<double> obj2 = stackalloc double[2] { 1, 1.2 };
```
Comment (VSadov) [inline src/Compilers/CSharp/Test/Semantic/Semantics/StackAllocInitializerTests.cs]: Can we have one more test similar to this with self-referential initializers? 
Like 
`Span<double> obj2 = stackalloc double[2] { obj2[0], obj2.Length };`
`double* obj2 = stackalloc double[2] { obj2[0], *obj2};`

These are error cases, but the semantic model should not go into infinite loop or something like that.

Comment (alrz) [reply src/Compilers/CSharp/Test/Semantic/Semantics/StackAllocInitializerTests.cs]: @VSadov Added. Also, please review https://github.com/dotnet/roslyn/pull/25384 which contains fixups for merge.

File: src/Compilers/CSharp/Portable/BoundTree/Expression.cs [hunk 2]
```diff
@@ -106,12 +106,17 @@ internal partial class BoundFixedLocalCollectionInitializer
 
     internal partial class BoundStackAllocArrayCreation
     {
-        protected override ImmutableArray<BoundNode> Children => ImmutableArray.Create<BoundNode>(this.Count);
+        internal static ImmutableArray<BoundExpression> GetChildInitializers(BoundArrayInitialization arrayInitializer)
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/BoundTree/Expression.cs]: private?

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_StackAllocArrayCreationAndInitializer.cs [hunk 3]
```diff
@@ -0,0 +1,705 @@
...
+            };
+
+            VerifyOperationTreeAndDiagnosticsForTest<StackAllocArrayCreationExpressionSyntax>(source, expectedOperationTree, expectedDiagnostics);
+        }
+
+        [Fact]
+        public void SimpleStackAllocArrayCreation_NonConstantDimension()
+        {
+            string source = @"
+struct M { }
```
Comment (jcouv) [inline src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_StackAllocArrayCreationAndInitializer.cs]: Nit: in the future, consider naming `S` (more conventional)

File: src/Compilers/CSharp/Portable/BoundTree/Expression.cs [hunk 4]
```diff
@@ -106,12 +106,17 @@ internal partial class BoundFixedLocalCollectionInitializer
...
+        {
+            return arrayInitializer?.Initializers ?? ImmutableArray<BoundExpression>.Empty;
+        }
+
+        protected override ImmutableArray<BoundNode> Children => StaticCast<BoundNode>.From(GetChildInitializers(this.InitializerOpt).Insert(0, this.Count));
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/BoundTree/Expression.cs]: Are the `Children` properties covered by test?

File: src/Compilers/CSharp/Portable/BoundTree/Expression.cs (hunk 2 above)
Comment (alrz) [reply src/Compilers/CSharp/Portable/BoundTree/Expression.cs]: should be either `protected` or `internal`. we could use `private`, make `Children` sealed and remove the override from ConvertedStackAlloc, but I think that's not practiced here.

File: src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_StackAllocArrayCreationAndInitializer.cs (hunk 3 above)
Comment (alrz) [reply src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_StackAllocArrayCreationAndInitializer.cs]: (inherited from array creation tests)

File: src/Compilers/CSharp/Portable/BoundTree/Expression.cs (hunk 4 above)
Comment (alrz) [reply src/Compilers/CSharp/Portable/BoundTree/Expression.cs]: Not directly, but the elements appears in IOperation tests.

**Comment by VSadov** [comment]
@dotnet-bot test windows_debug_unit64_prtest please

**Pull request merged**